salt/modules/yumpkg.py

```diff
@@ -1535,10 +1539,13 @@ def install(name=None,
                     to_install.append((pkgname, pkgstr))
                 else:
                     for ver in cver:
-                        ver = norm_epoch(ver, version_num)
-                        if salt.utils.versions.compare(ver1=version_num,
+                        # if diff_attr was provided, version number is directly
+                        # in ver, otherwise its a dict.
+                        ver2 = ver['version'] if diff_attr else ver
+                        ver2 = norm_epoch(ver2, version_num)
```

Comment:
Pascal programmers? :wink: 
```python
ver = norm_epoch(ver['version'] if diff_attr else ver, version_num)
```

Reply:
I am fine if you move all that to another function that would "extract version" or "normalise version" etc, as long this is needed again somewhere.

Reply:
@brejoc as of "preventing from bugs" it doesn't, but only adds to. Because I see it as a broken logic right now: in a loop you have a variable that can be A-type and B-type, while the only consumer you have can take B-type. Typically, a Java programmer will right away create another variable, because they cannot swap a type. In Python we do not need to pollute the scope with such unnecessary stuff, since we can extract data and reset the pointer where we need to, then pass to the consumer.

As of "code will grow" — when it will grow, _then_ they will add. No need to make complexity where it is not needed.

Reply:
@isbm I think this is getting a little bit too far - it's been 10 comments now just to argue about a one-liner versus the current 2 lines that add one extra (immutable) variable.

I suggest to call this a style difference, accept the fact different people have different styles, and just end the discussion here.

:peace_symbol:

Reply:
@moio there is no war, so :peace_symbol: is misleading here. Merging this and showing this as a bad example to others who will copy such things "because others did this already" — is a bad idea.

I am down-voting this multiline Java coding, but if @terminalmage says "yeah OK", then let it be so.

Reply:
@isbm There is something we can agree on. I also don't like the fact that `ver` can either be a string or a dict. So how about a little bit of refactoring to make sure that `cver` is always a list of dicts, even if `version` is the only key?

Reply:
@brejoc correction: _tuple_ or a dict. :wink: I absolutely for that refactoring then. Agreed. :stuck_out_tongue_winking_eye:

Reply:
Here is the [branch](https://github.com/brejoc/salt/commits/develop-yum-install-refactoring) with the refactoring.